RA-1888: Mark Patient Deceased

insookwa opened the pull request:
I have made the recommended changes.

https://issues.openmrs.org/browse/RA-1888

kindly review please @dkayiwa  @kdaud  @sherrif10

HerbertYiga commented:
thanks @insookwa  can you also add the ticket id here?

kdaud commented on ui-tests/src/test/java/org/openmrs/reference/page/MarkPatientDeceasedPage.java:
```diff
@@ -0,0 +1,127 @@
+	private static final By ACTIVE_DAY = By.cssSelector("td.day.active");
+	private static final By DEATH_MESSAGE = By.className("death-message");
+	private static final By EDIT = By.id("editConcept");
+	private static final By DELETE_BUTTON = By.xpath("//*[@id=\"saveDeleteButtons\"]/input[4]");
```
Comment for code line from `12` to `36`
Retrieving page elements through `xPath` especially via` x-value` tend to get broken in the long run when UI changes comes in. Use rather `id()` or  `cssSelector()` or `className()`  so that we opt for attributes that can not easily break and can support code substantiality in the long run.

kdaud commented:
@insookwa make sure all your files are well indented. See one of existing file indention [here](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/master/ui-tests/src/test/java/org/openmrs/reference/MergeVisitsTest.java) compare and contrast with yours [here ](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/35a27389514ba753a18d2a08ccc908781fd8fbb4/ui-tests/src/test/java/org/openmrs/reference/MarkPatientDeceasedTest.java) and see the difference then figure out to make the indention similar to the  existing one.

kdaud commented:
Great work @insookwa 
Catch my comments above!

sherrif10 commented on ui-tests/src/test/java/org/openmrs/reference/page/MarkPatientDeceasedPage.java:
```diff
@@ -0,0 +1,127 @@
+	private static final By SAVE = By.cssSelector("#saveDeleteButtons > input:nth-child(1)");
+	private static final By SEARCH = By.id("inputNode");
+	private static final By CONCEPT_SEARCH_MESSAGE = By.id("openmrsSearchTable_info");
+	private static final By COD = By.xpath("//*[@id=\"openmrsSearchTable\"]/tbody/tr[1]/td/span");
```
Is this COD , or CODE? , it need to be more readable. Thanks.

insookwa replied:
stands for Cause Of Death

insookwa replied:
Thanks, @kdaud  and @sherrif10  let me work on the changes

insookwa commented:
made some changes on this kindly review @kdaud @sherrif10

sherrif10 commented:
Hello @dkayiwa  Requesting for github approval

sherrif10 commented on ui-tests/src/test/java/org/openmrs/reference/MarkPatientDeceasedTest.java:
```diff
@@ -0,0 +1,45 @@
+	@Category(BuildTests.class)
+	public void markPatientDeceasedTest() {
+		homePage.goToAdministration();
+		MarkPatientDeceasedPage markPatientDeceased = new MarkPatientDeceasedPage(page);
```
Would you create the new instance of markPatientDeceasedPage in the before method. and just call the reference

sherrif10 commented on ui-tests/src/test/java/org/openmrs/reference/MarkPatientDeceasedTest.java:
```diff
@@ -0,0 +1,45 @@
+		findPatientPage.enterPatient(patient.identifier);
+		findPatientPage.waitForPageToLoad();
+		findPatientPage.clickOnFirstPatient();
+		markPatientDeceased.markPatientDead();
```
This is nice, However, would you mind calling is as **clickOnMarkPatientDead()**.

sherrif10 commented on ui-tests/src/test/java/org/openmrs/reference/page/MarkPatientDeceasedPage.java:
```diff
@@ -0,0 +1,114 @@
+		getConceptid(newId());
+	}
+	
+	public void markPatientDead() {
```
**clickOnMarkPatientDead() would suffice** . Thanks @insookwa .

sherrif10 commented on ui-tests/src/test/java/org/openmrs/reference/page/MarkPatientDeceasedPage.java:
```diff
@@ -0,0 +1,114 @@
+	private static final By ADD_ANSWER= By.cssSelector("#codedDatatypeRow > td > table > tbody > tr > td.buttons > input:nth-child(1)");
+	private static final By CAUSEBOX = By.id("newAnswerConcept");
+	private static final By FEVER = By.xpath("/html/body/ul[1]/li[2]/a");
+	private static final By ADDBUTTON = By.xpath("/html/body/div[5]/div[11]/button[1]");
```
Making it  as ADD_BUTTON  would look so nice .Thank.

sherrif10 commented:
Thanks @insookwa  this looks so awesome, Have left some minor comments.

sherrif10 commented on ui-tests/src/test/java/org/openmrs/reference/page/MarkPatientDeceasedPage.java:
```diff
@@ -0,0 +1,114 @@
+	private static final By SEARCH = By.id("inputNode");
+	private static final By COD = By.xpath("//*[@id=\"openmrsSearchTable\"]/tbody/tr[1]/td/span");
+	private static final By NEW_COD_ID = By.xpath("//*[@id=\"conceptTable\"]/tbody/tr[1]/td");
+	private static final By ADMINISTRATION = By.id("administrationNavLink");
```
Line 22 .is this administrationPage or administration link.

insookwa replied:
`<li id="administrationNavLink">
			<a href="/openmrs/admin/index.htm">Administration</a>
		</li>`

sherrif10 replied:
Do you mind sharing a  url link to this where you are inspecting from please, want to confirm .Thanks

insookwa replied:
https://qa-refapp.openmrs.org/openmrs/dictionary/index.htm   right click on administration and inspect element

sherrif10 replied:
> https://qa-refapp.openmrs.org/openmrs/dictionary/index.htm right click on administration and inspect element

Thanks , since administration link you have provided returns the same thing with administrationPage from this link https://github.com/openmrs/openmrs-distro-referenceapplication/blob/master/ui-tests/src/test/java/org/openmrs/reference/page/AdministrationPage.java  , why not simply using this method

kdaud commented:
@insookwa please make sure that your code indention aligns with the existing indention for quality assurance. Have you observed the diffrence between your file [indention](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/3bebdb73f94697df9943c7f282ca309ef470875e/ui-tests/src/test/java/org/openmrs/reference/MarkPatientDeceasedTest.java) and the exiting file indention for example [MergeVisitTest](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/master/ui-tests/src/test/java/org/openmrs/reference/MergeVisitsTest.java) ? Compare  and contrast !

kdaud commented:
@insookwa  you may need to revisit my comment [here](https://github.com/openmrs/openmrs-distro-referenceapplication/pull/529#issuecomment-899374814)!

insookwa commented:
> @insookwa you may need to revisit my comment [here](https://github.com/openmrs/openmrs-distro-referenceapplication/pull/529#issuecomment-899374814)!

is there an automatic way of fixing the indention issue

kdaud commented:
> is there an automatic way of fixing the indention issue

Do you mind googling  to figure that out ?

sherrif10 commented:
> is there an automatic way of fixing the indention issue

Sorry about that, Press CNTR +SHIFT + F , however you need to change your identation to 4 spaces , This might be of help https://wiki.openmrs.org/display/docs/Developer+How-To+Setup+And+Use+Eclipse

insookwa commented:
Thanks, @sherrif10

kdaud commented:
@insookwa please make sure all your files are well indented to correspond with the existing file indention.

kdaud commented:
@insookwa have you had a look at my  comment [above](https://github.com/openmrs/openmrs-distro-referenceapplication/pull/529#issuecomment-910550135) ?

insookwa commented:
> @insookwa have you had a look at my comment [above](https://github.com/openmrs/openmrs-distro-referenceapplication/pull/529#issuecomment-910550135) ?

hello @kdaud I was trying to copy from the reference you gave me

kdaud commented:
Figure out how best you can achieve the indention within your file.

insookwa commented:
> @insookwa have you had a look at my comment [above](https://github.com/openmrs/openmrs-distro-referenceapplication/pull/529#issuecomment-910550135) ?

Hello @kdaud  i have used the reference file for the required indention and it shows 4 which is the setting in my editor too .

insookwa commented:
I have made changes in the indention as advised. kindly review @kdaud 
@sherrif10

sherrif10 commented on ui-tests/src/test/resources/org/openmrs/uitestframework/test.properties:
```diff
@@ -1,4 +1,4 @@
-webapp.url=https://qa-refapp.openmrs.org/openmrs
+webapp.url=https://qa-refapp.openmrs.org/openmrs/
```
Thanks @insookwa . This looks good, However you can remove the slash at the end of this url.

insookwa replied:
i have removed it @sherrif10  thank you

sherrif10 commented:
One las thing , Is this fine to literally squash the commits to atleast less than 5 commits.

insookwa commented:
> One last thing , Is this fine to literally squash the commits to atleast less than 5 commits.

how about deleting the pull request /branch  and make a new clean one

kdaud commented on ui-tests/src/test/java/org/openmrs/reference/page/MarkPatientDeceasedPage.java:
```diff
@@ -0,0 +1,118 @@
+    private static final By ADD_NEW_CONCEPT = By.linkText("Add new Concept");
+    private static final By CONCEPT_NAME = By.id("namesByLocale[en].name");
+    private static final By ADD_ANSWER = By
+            .cssSelector("#codedDatatypeRow > td > table > tbody > tr > td.buttons > input:nth-child(1)");
```
Make the variable `ADD_ANSWER` be on one code line

kdaud commented on ui-tests/src/test/java/org/openmrs/reference/page/MarkPatientDeceasedPage.java:
```diff
@@ -0,0 +1,118 @@
+    private static final By CHECK_BOX = By.id("deceased");
+    private static final By DATE_PICKER = By.id("death-date-display");
+    private static final By SAVE_DECEASED = By
+            .cssSelector("#mark-patient-dead > fieldset > p:nth-child(4) > span:nth-child(2) > input");
```
The same with SAVE_DECEASED

kdaud commented on ui-tests/src/test/java/org/openmrs/reference/page/MarkPatientDeceasedPage.java:
```diff
@@ -0,0 +1,118 @@
+        findElement(SAVE_DECEASED).click();
+    }
+
+    public void newConcept() {
```
Rename the method to a  more readable name!

kdaud replied:
@insookwa did you forget to catch the previous [comment ](https://github.com/openmrs/openmrs-distro-referenceapplication/pull/529/files#r701124946) !

insookwa commented on ui-tests/src/test/java/org/openmrs/reference/page/MarkPatientDeceasedPage.java:
```diff
@@ -0,0 +1,118 @@
+        findElement(ACTIVE_DAY).click();
+    }
+
+    public String newId() {
```
Thanks @kdaud  i have changed them

kdaud commented on ui-tests/src/test/java/org/openmrs/reference/page/MarkPatientDeceasedPage.java:
```diff
@@ -0,0 +1,116 @@
+    private static final By SAVE_DECEASED = By.cssSelector("#mark-patient-dead > fieldset > p:nth-child(4) > span:nth-child(2) > input");
+    private static final By ACTIVE_DAY = By.cssSelector("td.day.active");
+    private static final By DEATH_MESSAGE = By.className("death-message");
+    private static final By EDIT = By.id("editConcept");
```
Are these variables from code line `12` to `34 ` buttons or Icons or links ?

insookwa replied:
some are buttons. some are links, others are messages and others are dates.

kdaud replied:
Its better to have readable names like EDIT_LINK, ACTIVE_DAY_BUTTON, DEATH_MESSAGE_FILED and so on

insookwa replied:
@kdaud  i have made some changes in relation to that comment . Thank you

sherrif10 commented:
Well done @insookwa , this is greatly improved.

kdaud commented on ui-tests/src/test/java/org/openmrs/reference/MarkPatientDeceasedTest.java:
```diff
@@ -0,0 +1,49 @@
+        assertTrue(markPatientDeceased.confirmDeadMessage().contains("The patient is deceased"));
+        homePage.go();
+        homePage.goToAdministration();
+        markPatientDeceased.deleteConcept();
```
Add an assertion method that affirms that the concept is deleted from the db

insookwa replied:
i added this

kdaud commented on ui-tests/src/test/java/org/openmrs/reference/page/MarkPatientDeceasedPage.java:
```diff
@@ -0,0 +1,116 @@
+    private static final By DICTIONARY_LINK= By.id("dictionaryNavLink");
+    private static final By ADD_NEW_CONCEPT = By.linkText("Add new Concept");
+    private static final By CONCEPT_NAME = By.id("namesByLocale[en].name");
+    private static final By ADD_ANSWER_BUTTON = By.cssSelector("#codedDatatypeRow > td > table > tbody > tr > td.buttons > input:nth-child(1)");
```
Are you able to get a better element look up for ADD_ANSWER_BUTTON ?

insookwa replied:
the ony availble ootion for cssselector is xpath . can i use that @kdaud

kdaud replied:
Does using `cssSelector("#codedDatatypeRow td table tbody tr td.buttons input:nth-child(1)");` retrieves the element ?

kdaud commented on ui-tests/src/test/java/org/openmrs/reference/page/MarkPatientDeceasedPage.java:
```diff
@@ -0,0 +1,116 @@
+    private static final By CAUSE_BOX = By.id("newAnswerConcept");
+    private static final By FEVER = By.xpath("/html/body/ul[1]/li[2]/a");
+    private static final By ADD_BUTTON = By.xpath("/html/body/div[5]/div[11]/button[1]");
+    private static final By SAVE_CONCEPT_BUTTON = By.xpath("/html/body/div[1]/div[3]/div[2]/form/div/input[1]");
```
Get better element reference for FEVER, ADD_BUTTON, and SAVE_CONCEPT_BUTTON rather than the xPath(). You can use` id()` or `cssSelector()`

insookwa replied:
i have removed all xpaths and used eith cssselector or id

kdaud commented on ui-tests/src/test/java/org/openmrs/reference/page/MarkPatientDeceasedPage.java:
```diff
@@ -0,0 +1,116 @@
+    private static final By CONCEPT_NAME = By.id("namesByLocale[en].name");
+    private static final By ADD_ANSWER_BUTTON = By.cssSelector("#codedDatatypeRow > td > table > tbody > tr > td.buttons > input:nth-child(1)");
+    private static final By CAUSE_BOX = By.id("newAnswerConcept");
+    private static final By FEVER = By.xpath("/html/body/ul[1]/li[2]/a");
```
Is FEVER a button or Icon or link or filed?

insookwa replied:
i changed this . thanks @kdaud

kdaud commented on ui-tests/src/test/java/org/openmrs/reference/page/MarkPatientDeceasedPage.java:
```diff
@@ -0,0 +1,116 @@
+    private static final By ADD_BUTTON = By.xpath("/html/body/div[5]/div[11]/button[1]");
+    private static final By SAVE_CONCEPT_BUTTON = By.xpath("/html/body/div[1]/div[3]/div[2]/form/div/input[1]");
+    private static final By SEARCH_CONCEPT_FIELD = By.id("inputNode");
+    private static final By FIRST_RESULT = By.xpath("//*[@id=\"openmrsSearchTable\"]/tbody/tr[1]/td/span");
```
What if you used `By.xpath("//*[@id='openmrsSearchTable']/tbody/tr[1]/td/span");` wouldn't it look better ?

As well comply FIRST_RESULT to my review a[bove](https://github.com/openmrs/openmrs-distro-referenceapplication/pull/529/files#r701711656) concerning variable name.

kdaud commented on ui-tests/src/test/java/org/openmrs/reference/page/MarkPatientDeceasedPage.java:
```diff
@@ -0,0 +1,116 @@
+    private static final By ADMINISTRATION_PAGE = By.id("administrationNavLink");
+    private static final By ADVANCED_SETTINGS = By.linkText("Advanced Settings");
+    private static final By CONCEPT_VALUE_FIELD = By.xpath("//*[@id=\"globalPropsList\"]/tr[133]/td[2]/input");
+    private static final By SAVE_CONCEPT_VALUE = By.xpath("//*[@id=\"buttonsAtBottom\"]/input[1]");
```
For CONCEPT_VALUE_FILED and SAVE_CONCEPT_VALUE element reference refer to my first paragraph [above](https://github.com/openmrs/openmrs-distro-referenceapplication/pull/529/files#r701714914)

insookwa replied:
i changed al these to either id or cssselector

kdaud commented on ui-tests/src/test/java/org/openmrs/reference/page/MarkPatientDeceasedPage.java:
```diff
@@ -0,0 +1,116 @@
+    private static final By SAVE_CONCEPT_VALUE = By.xpath("//*[@id=\"buttonsAtBottom\"]/input[1]");
+    private static final By HOME_LINK = By.id("homeNavLink");
+    private static final By MARK_PATIENT_DECEASED = By.id("org.openmrs.module.coreapps.markPatientDead");
+    private static final By CHECK_BOX = By.id("deceased");
```
Use a better element readable name rather than CHECK_BOX. 

Hint: CHECK_BOX for what ?

kdaud commented on ui-tests/src/test/java/org/openmrs/reference/page/MarkPatientDeceasedPage.java:
```diff
@@ -0,0 +1,116 @@
+    private static final By DATE_PICKER = By.id("death-date-display");
+    private static final By SAVE_DECEASED_BUTTON= By.cssSelector("#mark-patient-dead > fieldset > p:nth-child(4) > span:nth-child(2) > input");
+    private static final By ACTIVE_DAY = By.cssSelector("td.day.active");
+    private static final By DEATH_MESSAGE = By.className("death-message");
```
For ACTIVE_DAY, DEATH_MESSAGE refer to my to my comment [above](https://github.com/openmrs/openmrs-distro-referenceapplication/pull/529/files#r701711656)

kdaud commented on ui-tests/src/test/java/org/openmrs/reference/page/MarkPatientDeceasedPage.java:
```diff
@@ -0,0 +1,116 @@
+    private static final By SAVE_DECEASED_BUTTON= By.cssSelector("#mark-patient-dead > fieldset > p:nth-child(4) > span:nth-child(2) > input");
+    private static final By ACTIVE_DAY = By.cssSelector("td.day.active");
+    private static final By DEATH_MESSAGE = By.className("death-message");
+    private static final By EDIT_CONCEPT = By.id("editConcept");
```
For ACTIVE_DAY, DEATH_MESSAGE and EDT_CONCEPT refer to my comment [above](https://github.com/openmrs/openmrs-distro-referenceapplication/pull/529/files#r701711656) !

kdaud commented on ui-tests/src/test/java/org/openmrs/reference/page/MarkPatientDeceasedPage.java:
```diff
@@ -0,0 +1,116 @@
+    private static final By MARK_PATIENT_DECEASED = By.id("org.openmrs.module.coreapps.markPatientDead");
+    private static final By CHECK_BOX = By.id("deceased");
+    private static final By DATE_PICKER = By.id("death-date-display");
+    private static final By SAVE_DECEASED_BUTTON= By.cssSelector("#mark-patient-dead > fieldset > p:nth-child(4) > span:nth-child(2) > input");
```
Get a better element reference for SAVE_DECEASED_BUTTON that can hardly break !

insookwa replied:
i have changedd this . thanks @kdaud

kdaud replied:
> i have changedd this

I can still see the element reference for  [SAVE_DECEASED_BUTTON](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/2b3a370fd7297b5004048b226641583364517c42/ui-tests/src/test/java/org/openmrs/reference/page/MarkPatientDeceasedPage.java#L34) is not a good one in retrieving the element, it can easily break. Get a better one

kdaud commented:
@insookwa are you working on this PR ?

 For any help you may need am available or you can as well reach out to @sherrif10 !

insookwa commented:
> @insookwa are you working on this PR ?
> 
> For any help you may need am available or you can as well reach out to @sherrif10 !

hi @kdaud   the chrome tests are passing very well but firefox is failing with this :

	org.openqa.selenium.ElementClickInterceptedException: Element <input name="action" type="submit"> is not clickable at point (55,670) because another element <div class="ui-widget-overlay"> obscures it
Build info: version: '4.0.0-alpha-6', revision: '5f43a29cfc'

kdaud commented:
@insookwa  can you share a full log report at https://pastebin.com/  ?

insookwa commented:
> @insookwa always use [pastebin.com](https://pastebin.com/) to share a full log report. BTW have you addressed all the reviews on this PR before we look into the logs ?
Hello @kdaud  these are the logs 
https://pastebin.com/FLPArXjZ

kdaud commented:
@insookwa did you read my comment [above](https://github.com/openmrs/openmrs-distro-referenceapplication/pull/529#issuecomment-916305548) very well ?

insookwa commented:
> @insookwa always use [pastebin.com](https://pastebin.com/) to share a full log report. BTW have you addressed all the reviews on this PR before we look into the logs?

@kdaud  i have fixed all that was in the comments as advised on the last PRs. however, the test passes on local instance in chrome and firefox but firefox fails on online instance

insookwa commented:
looks like https://qa-refapp.openmrs.org/openmrs/ is down. could it e the cause of this test failure. @kdaud  @sherrif10

kdaud commented:
> looks like https://qa-refapp.openmrs.org/openmrs/ is down. could it e the cause of this test failure.

The server is up!

@insookwa am seeing still pending reviews that you have not yet addressed ?

 Would like to look into the logs when all the reviews have been resolved !

kdaud commented on ui-tests/src/test/java/org/openmrs/reference/page/MarkPatientDeceasedPage.java:
```diff
@@ -0,0 +1,123 @@
+    private static final By CAUSE_SEARCH_FIELD = By.id("newAnswerConcept");
+    private static final By FIRST_CAUSE_OPTION = By.className("autocompleteresult");
+    private static final By ADD_BUTTON = By.className("ui-button-text");
+    private static final By SAVE_CONCEPT_BUTTON = By.cssSelector("#saveDeleteButtons > input[type=submit]:nth-child(1)");
```
How about using  `By.cssSelector("#saveDeleteButtons input[type=submit]:nth-child(1)");` ?

insookwa replied:
this doesn't locate the element

kdaud commented on ui-tests/src/test/java/org/openmrs/reference/page/MarkPatientDeceasedPage.java:
```diff
@@ -0,0 +1,123 @@
+    private static final By SAVE_CONCEPT_BUTTON = By.cssSelector("#saveDeleteButtons > input[type=submit]:nth-child(1)");
+    private static final By SEARCH_CONCEPT_FIELD = By.id("inputNode");
+    private static final By FIRST_RESULT = By.className("odd");
+    private static final By CONCEPT_ID = By.cssSelector("#conceptTable > tbody > tr:nth-child(1) > td");
```
Does using `By.cssSelector("#conceptTable tbody tr:nth-child(1) td");` retrieves the the element ?

insookwa replied:
yes it does . it picks the concept id

kdaud replied:
> yes it does . it picks the concept id

Do the needful !

insookwa replied:
am sorry but this also breaks

kdaud commented on ui-tests/src/test/java/org/openmrs/reference/page/MarkPatientDeceasedPage.java:
```diff
@@ -0,0 +1,123 @@
+    private static final By CONCEPT_ID = By.cssSelector("#conceptTable > tbody > tr:nth-child(1) > td");
+    private static final By ADMINISTRATION_PAGE = By.id("administrationNavLink");
+    private static final By ADVANCED_SETTINGS = By.linkText("Advanced Settings");
+    private static final By CONCEPT_VALUE_FIELD = By.cssSelector("#globalPropsList > tr:nth-child(133) > td:nth-child(2) > input[type=text]");
```
Get a better element reference for CONCEPT_VALUE_FILED otherwise the current one can easily break !

kdaud commented on ui-tests/src/test/java/org/openmrs/reference/page/MarkPatientDeceasedPage.java:
```diff
@@ -0,0 +1,123 @@
+    private static final By ACTIVE_DAY = By.cssSelector("td.day.active");
+    private static final By DEATH_MESSAGE_TEXT = By.className("death-message");
+    private static final By EDIT_CONCEPT_LINK = By.id("editConcept");
+    private static final By DELETE_BUTTON = By.cssSelector("#saveDeleteButtons > input[type=submit]:nth-child(4)");
```
Does using `By.cssSelector("#saveDeleteButtons input[type=submit]:nth-child(4)");` retrieves the element ?

insookwa replied:
@kdaud  yes it does

kdaud replied:
Do the needfull

insookwa replied:
@kdaud  actually By.cssSelector("#saveDeleteButtons input[type=submit]:nth-child(4)"); doesnt retrive the element

kdaud commented on ui-tests/src/test/java/org/openmrs/reference/page/MarkPatientDeceasedPage.java:
```diff
@@ -0,0 +1,123 @@
+    }
+
+    public void clickOnMarkPatientDead() {
+        findElement(MARK_PATIENT_DECEASED_LINK).click();
```
Use rather `clickOn(elementName); `function than using `findElement(elementName).click();`. Check through all the file to replace findElement().click() with `clickOn();`

kdaud commented:
Great progress @insookwa 

Catch my comments above !

kdaud commented on ui-tests/src/test/java/org/openmrs/reference/page/MarkPatientDeceasedPage.java:
```diff
@@ -0,0 +1,123 @@
+     * if there is no concept at all or no concept called Cause of death, the method
+     * creates one and copies its concept code
+     */
+    
```
Remove code line `56`

kdaud replied:
@insookwa did you see my review [above](https://github.com/openmrs/openmrs-distro-referenceapplication/pull/529/files#r706739042) ?

kdaud commented:
@insookwa did you take a look at  my review [here](https://github.com/openmrs/openmrs-distro-referenceapplication/pull/529#discussion_r705953136) ?

kdaud commented on ui-tests/src/test/java/org/openmrs/reference/page/MarkPatientDeceasedPage.java:
```diff
@@ -0,0 +1,124 @@
+
+    private static final String URL = "/coreapps/markPatientDead.page";
+    private static final By DICTIONARY_LINK= By.id("dictionaryNavLink");
+    private static final By ADD_NEW_CONCEPT = By.linkText("Add new Concept");
```
is ADD_NEW_CONCEPT a button or a link or a filed ? Also retrieving an element text is via linkText() is not a good deal because they can easily break, opt for id() or cssSelcetor()

kdaud replied:
What about via xPath() ?

insookwa replied:
xpath breaks the test at some point

kdaud commented on ui-tests/src/test/java/org/openmrs/reference/page/MarkPatientDeceasedPage.java:
```diff
@@ -0,0 +1,124 @@
+    private static final By CONCEPT_NAME = By.id("namesByLocale[en].name");
+    private static final By ADD_ANSWER_BUTTON = By.cssSelector("#codedDatatypeRow > td > table > tbody > tr > td.buttons > input:nth-child(1)");
+    private static final By CAUSE_SEARCH_FIELD = By.id("newAnswerConcept");
+    private static final By FIRST_CAUSE_OPTION = By.className("autocompleteresult");
```
is FIRST_CAUSE_OPTION a button or what ?

insookwa replied:
this is the first option that appears when the search for the causes is done

kdaud replied:
@insookwa let me change the wording. is `FIRST_CAUSE_OPTION` a button or field or link ?

kdaud commented on ui-tests/src/test/java/org/openmrs/reference/page/MarkPatientDeceasedPage.java:
```diff
@@ -0,0 +1,124 @@
+    private static final By ADD_BUTTON = By.className("ui-button-text");
+    private static final By SAVE_CONCEPT_BUTTON = By.cssSelector("#saveDeleteButtons > input[type=submit]:nth-child(1)");
+    private static final By SEARCH_CONCEPT_FIELD = By.id("inputNode");
+    private static final By FIRST_RESULT = By.className("odd");
```
is FIRST_RESULT a button or what ?

kdaud commented on ui-tests/src/test/java/org/openmrs/reference/page/MarkPatientDeceasedPage.java:
```diff
@@ -0,0 +1,124 @@
+    private static final By SAVE_CONCEPT_BUTTON = By.cssSelector("#saveDeleteButtons > input[type=submit]:nth-child(1)");
+    private static final By SEARCH_CONCEPT_FIELD = By.id("inputNode");
+    private static final By FIRST_RESULT = By.className("odd");
+    private static final By CONCEPT_ID = By.cssSelector("#conceptTable > tbody > tr:nth-child(1) > td");
```
is CONCEPT_ID a button, link or what ?

insookwa replied:
i fixed these

kdaud commented on ui-tests/src/test/java/org/openmrs/reference/page/MarkPatientDeceasedPage.java:
```diff
@@ -0,0 +1,124 @@
+    private static final By FIRST_RESULT = By.className("odd");
+    private static final By CONCEPT_ID = By.cssSelector("#conceptTable > tbody > tr:nth-child(1) > td");
+    private static final By ADMINISTRATION_PAGE = By.id("administrationNavLink");
+    private static final By ADVANCED_SETTINGS = By.linkText("Advanced Settings");
```
is ADVANCED_SETTINGS a button, link or what ?

insookwa replied:
i fixed these

kdaud commented on ui-tests/src/test/java/org/openmrs/reference/page/MarkPatientDeceasedPage.java:
```diff
@@ -0,0 +1,124 @@
+    private static final By HOME_LINK = By.id("homeNavLink");
+    private static final By MARK_PATIENT_DECEASED_LINK = By.id("org.openmrs.module.coreapps.markPatientDead");
+    private static final By MARK_PATIENT_DECEASED_CHECK_BOX = By.id("deceased");
+    private static final By DATE_PICKER = By.id("death-date-display");
```
even here show clearly whether the element is a button, link or something

insookwa replied:
i have fixed these

kdaud commented on ui-tests/src/test/java/org/openmrs/reference/page/MarkPatientDeceasedPage.java:
```diff
@@ -0,0 +1,124 @@
+    private static final By MARK_PATIENT_DECEASED_CHECK_BOX = By.id("deceased");
+    private static final By DATE_PICKER = By.id("death-date-display");
+    private static final By SAVE_DECEASED_BUTTON= By.cssSelector("#mark-patient-dead > fieldset > p:nth-child(4) > span:nth-child(2) > input");
+    private static final By ACTIVE_DAY = By.cssSelector("td.day.active");
```
even here !

insookwa replied:
i have fixed this

kdaud commented on ui-tests/src/test/java/org/openmrs/reference/page/MarkPatientDeceasedPage.java:
```diff
@@ -0,0 +1,124 @@
+        clickOn(DICTIONARY_LINK); 
+        Concept();
+        clickOn(HOME_LINK); 
+     
```
Remove code line `48`

insookwa replied:
i have fixed this

kdaud commented on ui-tests/src/test/java/org/openmrs/reference/page/MarkPatientDeceasedPage.java:
```diff
@@ -0,0 +1,124 @@
+        clickOn(MARK_PATIENT_DECEASED_CHECK_BOX);
+        clickOn(DATE_PICKER);
+        pickDate();
+        Select dropdown = new Select(findElement(By.id("cause-of-death")));
```
Get a better way to select the element from the drop down list

insookwa replied:
Thanks @kdaud  do you have any recommended way i can refer to

kdaud commented on ui-tests/src/test/java/org/openmrs/reference/page/MarkPatientDeceasedPage.java:
```diff
@@ -0,0 +1,124 @@
+        pickDate();
+        Select dropdown = new Select(findElement(By.id("cause-of-death")));
+        dropdown.selectByValue("113230AAAAAAAAAAAAAAAAAAAAAAAAAAAAAA");
+        findElement(SAVE_DECEASED_BUTTON).click();
```
What about using `clickOn(SAVE_DECEASED_BUTTON)` ?

insookwa replied:
i have fixed this

kdaud commented on ui-tests/src/test/java/org/openmrs/reference/page/MarkPatientDeceasedPage.java:
```diff
@@ -0,0 +1,124 @@
+    public void newConcept() {
+        clickOn(ADD_NEW_CONCEPT); 
+        findElement(CONCEPT_NAME).sendKeys("cause of death");
+        Select datatype = new Select(findElement(By.id("datatype")));
```
Get a better way to select the datatype

kdaud commented on ui-tests/src/test/java/org/openmrs/reference/page/MarkPatientDeceasedPage.java:
```diff
@@ -0,0 +1,124 @@
+        clickOn(ADD_BUTTON);
+        waiter.until(ExpectedConditions.elementToBeClickable(SAVE_CONCEPT_BUTTON));
+        clickOn(SAVE_CONCEPT_BUTTON); 
+
```
Remove code line `83`

insookwa replied:
Hello @kdaud  i have worked on all the reviews  kindly review

kdaud commented:
Great work @insookwa 

Have left few comments for you !

sherrif10 commented:
@insookwa  Hello Happy seeing your input again, Lets have pages written in qaframework for consistency. Thanks

sherrif10 commented:
Hey @insookwa  we can close this since we shall be catering this in qaframework module.Thanks.

kdaud commented:
Closing this PR since selenium legacy tests were migrated to [qaframework-repo](https://github.com/openmrs/openmrs-contrib-qaframework/tree/master/qaframework-legacy-tests/src/test/java/org/openmrs/contrib/qaframework/legacy) !!
